source/aws_http_credentials_provider.c

```diff
@@ -0,0 +1,515 @@
+            .secret_access_key_name = "secretAccessKey",
+            .token_name = "sessionToken",
+            .expiration_name = "expiration",
+            .top_level_object_name = "roleCredentials",
+            .token_required = true,
+            .expiration_required = true,
+            .expiration_format = AWS_PCEF_NUMBER_UNIX_EPOCH_MS,
+        };
+
+        credentials = aws_parse_credentials_from_json_document(
+            http_query_context->allocator, aws_byte_cursor_from_buf(&http_query_context->payload), &parse_options);
+    }
+
+    if (credentials) {
+        AWS_LOGF_INFO(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p) successfully queried credentials",
+            (void *)http_query_context->provider);
+        struct aws_credentials_options creds_option =
+            impl->request_vtable->create_credentials_options_fn(credentials, http_query_context);
+        credentials_with_account_id = aws_credentials_new_with_options(http_query_context->allocator, &creds_option);
+    } else {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p) failed to query credentials",
+            (void *)http_query_context->provider);
+
+        if (http_query_context->error_code == AWS_ERROR_SUCCESS) {
+            http_query_context->error_code = http_query_context->error;
+        }
+    }
+
+    /* pass the credentials back */
+    http_query_context->original_callback(
+        credentials_with_account_id, http_query_context->error_code, http_query_context->original_user_data);
+
+    /* clean up */
+    s_http_query_context_destroy(http_query_context);
+    aws_credentials_release(credentials);
+    aws_credentials_release(credentials_with_account_id);
+}
+
+static void s_on_retry_ready(struct aws_retry_token *token, int error_code, void *user_data);
+
+static void s_on_stream_complete_fn(struct aws_http_stream *stream, int error_code, void *user_data) {
+    struct aws_http_query_context *http_query_context = user_data;
+
+    struct aws_credentials_provider_http_impl *impl = http_query_context->provider->impl;
+    impl->function_table->aws_http_stream_release(stream);
+
+    /* set error code */
+    http_query_context->error_code = error_code;
+    impl->function_table->aws_http_stream_get_incoming_response_status(stream, &http_query_context->status_code);
+    if (error_code == AWS_OP_SUCCESS && http_query_context->status_code != AWS_HTTP_STATUS_CODE_200_OK) {
+        http_query_context->error_code = AWS_AUTH_CREDENTIALS_PROVIDER_HTTP_STATUS_FAILURE;
+    }
+
+    /*
+     * If we can retry the request based on error response or http status code failure, retry it, otherwise, call the
+     * finalize function.
+     */
+    if (error_code || http_query_context->status_code != AWS_HTTP_STATUS_CODE_200_OK) {
+        enum aws_retry_error_type error_type =
+            aws_credentials_provider_compute_retry_error_type(http_query_context->status_code, error_code);
```

Comment:
it's not really part of the PR, but the status code can be 0, in case we don't have the full response back, in which case, we will rely on the error code itself to determine it's retriable or not. 

But the current `aws_credentials_provider_compute_retry_error_type`, doesn't seem to handle the 0 status code case well, it will just fallback to `AWS_RETRY_ERROR_TYPE_SERVER_ERROR`. 

Either we just default the status code to something like 400, or we handle the case where status code is not set in the helper.

Reply:
> But the current aws_credentials_provider_compute_retry_error_type

I know this is not the best answer but i kind of want to avoid touching that for the sake of this PR because it will touch _all credentials provider_ and i want to reduce the radius of this change if possible. I am open to working this into this if you want it though